classes/notification/PKPNotificationOperationManager.inc.php

```diff
@@ -385,19 +386,87 @@ private function sendNotificationEmail($request, $notification) {
 				$mail->setReplyTo($site->getLocalizedContactEmail(), $site->getLocalizedContactName());
 			}
 
+			$unsubscribeUrl = $this->getUnsubscribeNotificationUrl($request, $notification);
+
 			$mail->assignParams(array(
 				'notificationContents' => $this->getNotificationContents($request, $notification),
 				'url' => $this->getNotificationUrl($request, $notification),
-				'siteTitle' => $context?$context->getLocalizedName():$site->getLocalizedTitle()
+				'siteTitle' => $context?$context->getLocalizedName():$site->getLocalizedTitle(),
+				'notificationUnsubscribeUrl' => $unsubscribeUrl,
 			));
+
 			$mail->addRecipient($user->getEmail(), $user->getFullName());
+
 			if (!$mail->send()) {
 				import('classes.notification.NotificationManager');
 				$notificationMgr = new NotificationManager();
 				$notificationMgr->createTrivialNotification($request->getUser()->getId(), NOTIFICATION_TYPE_ERROR, array('contents' => __('email.compose.error')));
 			}
 		}
 	}
+
+	/**
+	 * Creates and returns a unique string for the given notification, that will be encoded and validated against.
+	 * @param $notification Notification
+	 * @return string
+	 */
+	function createUnsubscribeUniqueKey($notification) {
+		$uniqueKey = 'unsubscribe' . '-' . $notification->getContextId() . '-' . $notification->getUserId() . '-' . $notification->getId();
```

Comment:
Is it possible to have a notification without a context id? Are there admin-only notifications?

Reply:
If without `contextId` means that `$contextId == null` then if I am not mistaken it's not possible. The `notifications` database table has the `context_id` column `not null`. 
If it means `$contextId == 0` it could be possible, but then wouldn't the `notification` have its `contextId == 0`?

Also the unsubscribe action is context related in the sense that a user can have different email notifications configuration in `context1` and `context2`. 

So, I suppose the `contextId` should probably be included.

Reply:
Sounds good. :+1: